ssl/record/rec_layer_s3.c

```diff
@@ -751,10 +751,29 @@ int do_ssl3_write(SSL *s, int type, const unsigned char *buf,
             size_t tmppipelen = 0;
             int ret;
 
-            ret = do_ssl3_write(s, type, buf, &tmppipelen, 1, 1, &prefix_len);
-            if (ret <= 0) {
-                /* SSLfatal() already called if appropriate */
-                goto err;
+            if (BIO_get_ktls_send(s->wbio)) {
+                /*
+                 * For KTLS, send an explicit empty fragment rather than
+                 * storing the fragment as a prefix in the buffer.
+                 */
+                ret = BIO_flush(s->wbio);
+                if (ret <= 0)
+                    goto err;
+
+                BIO_set_ktls_ctrl_msg(s->wbio, type);
+                /* TODO(size_t): Convert this call */
+                ret = BIO_write(s->wbio, NULL, 0);
+
+                /* Assume a zero byte write is success, not an error. */
```

Comment:
How do we distinguish a failed write (retry required) from success?

Reply:
With the BIO interface as-is, we cannot, hence the explicit comment about the assumption.   Even using BIO_write_ex() I think doesn't help as the implementations are the same as @kaduk pointed out.

Reply:
if `ret <= 0` I think we should call `BIO_should_retry()`. If that returns true then we should `goto err`. If it returns false then we assume `ret == 0` is success and `ret < 0` is failure (in which case set `s->rwstate = SSL_NOTHING` and then `goto err`). If we got 0 and it is a "real" non-retryable failure then we should fail on the next write anyway.